libraries/botbuilder-ai/src/luis-client/luisClient.ts

```diff
@@ -0,0 +1,301 @@
+/**
+ * @module botbuilder-ai
+ */
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+/**
+ * NOTE: This class was auto generated by OpenAPI Generator (https://openapi-generator.tech)
+ * and was modify manually to make compliance with the current implementation of the library.
+ */
+
+import localVarRequest = require('request');
+import http = require('http');
+
+/* tslint:disable:no-unused-locals */
+import { LuisResult } from './model/luisResult';
+
+import { ObjectSerializer, Authentication, VoidAuth } from './model/models';
+import { ApiKeyAuth } from './model/models';
+
+let luisVersion = '/luis/v2.0';
+
+export interface PredictionResolveOptionalParams {
+    /**
+     * The timezone offset for the location of the request.
+     */
+    timezoneOffset?: number;
+    /**
+     * If true, return all intents instead of just the top scoring intent.
+     */
+    verbose?: boolean;
+    /**
+     * Use the staging endpoint slot.
+     */
+    staging?: boolean;
+    /**
+     * Enable spell checking.
+     */
+    spellCheck?: boolean;
+    /**
+     * The subscription key to use when enabling bing spell check
+     */
+    bingSpellCheckSubscriptionKey?: string;
+    /**
+     * Log query (default is true)
+     */
+    log?: boolean;
+
+    customHeaders: {headers: {[name: string]: string}};
+}
+
+export enum LuisApikeys {
+    apiKeyHeader,
+}
+
+export class LuisClient {
+    protected _basePath: string = '';
+    protected defaultHeaders: any = {};
+    protected _useQuerystring: boolean = false;
+
+    protected authentications = {
+        'default': new VoidAuth() as Authentication,
+        'apiKeyHeader': new ApiKeyAuth('header', 'Ocp-Apim-Subscription-Key'),
+    }
+
+    public constructor(basePath: string){
+        if (basePath) {
+            this.basePath = basePath + luisVersion;
+        }
+    }
+
+    public set useQuerystring(value: boolean) {
+        this._useQuerystring = value;
+    }
+
+    public set basePath(basePath: string) {
+        this._basePath = basePath;
+    }
+
+    public get basePath(): string {
+        return this._basePath;
+    }
+
+    public setDefaultAuthentication(auth: Authentication): void {
+        this.authentications.default = auth;
+    }
+
+    public setApiKey(key: LuisApikeys, value: string): void {
+        (this.authentications as any)[LuisApikeys[key]].apiKey = value;
+    }
+
+    /**
+     * Gets predictions for a given utterance, in the form of intents and entities. The current maximum query size is 500 characters.
+     * @param query The utterance to predict.
+     * @param appId The LUIS application ID (Guid).
+     * @param [options] The optional parameters
+     * @returns Promise<LuisResult>
+     */
+    public async predictionResolvePost(query: string, appId: string, options: PredictionResolveOptionalParams): Promise<LuisResult> {
```

Comment:
**NOTE**: A possible improvement could be refactoring the `predictionResolvePost` and `predictionResolvePost` methods, as they share _almost_ the same code (actually, there is a lot of repeated code that could be refactored into new methods, as the validation part and the returned promise)

Reply:
I agree. The two methods can and should be refactored.